appcheck/appcheck.go

```diff
@@ -0,0 +1,136 @@
+// Package appcheck provides functionality for verifying App Check tokens.
+package appcheck
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/MicahParks/keyfunc"
+	"github.com/golang-jwt/jwt/v4"
+
+	"firebase.google.com/go/v4/internal"
+)
+
+// JWKSUrl is the URL of the JWKS used to verify App Check tokens.
+const JWKSUrl = "https://firebaseappcheck.googleapis.com/v1beta/jwks"
+
+const appCheckIssuer = "https://firebaseappcheck.googleapis.com/"
+
+var (
+	// ErrIncorrectAlgorithm is returned when the token is signed with a non-RSA256 algorithm.
+	ErrIncorrectAlgorithm = errors.New("token has incorrect algorithm")
+	// ErrTokenType is returned when the token is not a JWT.
+	ErrTokenType = errors.New("token has incorrect type")
+	// ErrTokenClaims is returned when the token claims cannot be decoded.
+	ErrTokenClaims = errors.New("token has incorrect claims")
+	// ErrTokenAudience is returned when the token audience does not match the current project.
+	ErrTokenAudience = errors.New("token has incorrect audience")
+	// ErrTokenIssuer is returned when the token issuer does not match Firebase's App Check service.
+	ErrTokenIssuer = errors.New("token has incorrect issuer")
+	// ErrTokenSubject is returned when the token subject is empty or missing.
+	ErrTokenSubject = errors.New("token has empty or missing subject")
+)
+
+// VerifiedToken represents a verified App Check token.
+type VerifiedToken struct {
+	Iss   string
+	Sub   string
+	Aud   []string
+	Exp   time.Time
+	Iat   time.Time
+	AppID string
+}
+
+// Client is the interface for the Firebase App Check service.
+type Client struct {
+	projectID string
+
+	jwks *keyfunc.JWKS
+}
+
+// NewClient creates a new App Check client.
+func NewClient(ctx context.Context, conf *internal.AppCheckConfig) (*Client, error) {
+	// TODO: Add support for overriding the HTTP client using the App one.
+	jwks, err := keyfunc.Get(conf.JWKSUrl, keyfunc.Options{
+		Ctx: ctx,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		projectID: conf.ProjectID,
+		jwks:      jwks,
+	}, nil
+}
+
+// VerifyToken verifies the given App Check token.
```

Comment:
```suggestion
// VerifyToken Verifies the given App Check token.
//
// VerifyToken considers an App Check token string to be valid if all the following conditions are met:
//   - The token string is a valid RS256 JWT.
//   - The JWT contains valid issuer (iss) and audience (aud) claims that match the issuerPrefix
//     and projectID of the tokenVerifier.
//   - The JWT contains a valid subject (sub) claim.
//   - The JWT is not expired, and it has been issued some time in the past.
//   - The JWT is signed by a Firebase App Check backend server as determined by the keySource.
//
// If any of the above conditions are not met, an error is returned. Otherwise a pointer to a
// decoded App Check token is returned.
```

Reply:
Done.